### DISCUSSION TITLE
TSDX with TailwindCSS PostCSS plugin -- `inject` and `purge` options

### DISCUSSION THREAD

**Discussion by leo-petrucci** [jaredpalmer/tsdx#947] [Q&A] [answered]
I'm new to tsdx and I'm having trouble setting up TailwindCSS. I've created a `tsdx.config.js` and set up the needed postCSS plugins:

```Javascript
const postcss = require('rollup-plugin-postcss');
const autoprefixer = require('autoprefixer');
const cssnano = require('cssnano');
const tailwindcss = require('tailwindcss');

module.exports = {
  rollup(config, options) {
    config.plugins.push(
      postcss({
        plugins: [
          tailwindcss({
            purge: ['./src/**/*.tsx'],
            darkMode: false, // or 'media' or 'class'
            theme: {
              extend: {},
            },
            variants: {
              extend: {},
            },
            plugins: [],
          }),
          autoprefixer(),
        ],
        inject: false,
        // only write out CSS for the first bundle (avoids pointless extra files):
        extract: !!options.writeMeta,
      })
    );
    return config;
  },
};
```

I then import my `src/styles.css` in one of my `.tsx` files inside `src`.

```
// styles.css
@tailwind base;
@tailwind components;
@tailwind utilities;
```

TSDX then actually does build the CSS file and outputs it correctly at `dist/modal.cjs.development.css` but:

- None of the css is purged
- The css isn't imported in any of my components, so no styles are applied

Any tips on getting this to work?

**Comment by leo-petrucci** [comment, marked as answer]
Looks like the problem was with: `inject`. I switched it to `inject: true,` and css is working. 

Purging can be fixed by changing the build command to `"build": "NODE_ENV=production tsdx build",`.

**Reply by agilgur5** [reply]
Yea [`inject`](https://github.com/egoist/rollup-plugin-postcss#inject) is set to `false` in [the example](https://github.com/formium/tsdx/tree/995099fd1d60104bca9ba334f8cd8c0835122840#example-adding-postcss) following the library pattern of providing a CSS file that your consumers can import. You can inject the styles yourself, but that can be a bit intrusive as a library author -- https://github.com/formium/tsdx/issues/186#issuecomment-616913315 and surrounding comments go into this in some detail.

This does seem to be specific to `rollup-plugin-postcss` and not really TSDX. Why Tailwind's `purge` option doesn't work is a question for the Tailwind PostCSS plugin too. There is [an integration test](https://github.com/formium/tsdx/blob/462af2d002987f985695b98400e0344b8f2754b7/test/integration/tsdx-build-withConfig.test.ts#L38) for `cssnano` and `autoprefixer`, so those work.

It sounds like based on your findings that Tailwind's plugin requires `NODE_ENV` to be set instead of having a configuration (might be a mismatch in Rollup vs PostCSS operating modes, though it's not mentioned in the Tailwind docs). There is an issue for setting `NODE_ENV` but for the Babel use-case instead of PostCSS use-case: https://github.com/formium/tsdx/issues/687 . It hasn't gotten any up-votes so it does not seem to be that often used/needed.
I'm not sure what other ramifications there are of setting `NODE_ENV` though or how that may impact other plugins

**Comment by stephencweiss** [comment]
**Update:** In addition to changing `inject`, you also need to ensure that `extract` is `false` because [`inject` is always false when `extract: true`](https://github.com/egoist/rollup-plugin-postcss#inject).

@creativiii - am I correct that once you modify the `inject` property to `true`, you don't need anything else? No `.postcssrc` in the `example/`, etc.?

```diff
const postcss = require('rollup-plugin-postcss');
const autoprefixer = require('autoprefixer');
const cssnano = require('cssnano');
const tailwindcss = require('tailwindcss');

module.exports = {
  rollup(config, options) {
    config.plugins.push(
      postcss({
        plugins: [
          tailwindcss({
            purge: ['./src/**/*.tsx'],
            darkMode: false, // or 'media' or 'class'
            theme: {
              extend: {},
            },
            variants: {
              extend: {},
            },
            plugins: [],
          }),
          autoprefixer(),
        ],
-          inject: false,
+          inject: true,
        // only write out CSS for the first bundle (avoids pointless extra files):
        extract: !!options.writeMeta,
      })
    );
    return config;
  },
};
```

**Comment by hoangvu12** [comment]
Is there anyway to add TailwindCSS to example? I just want to use tailwind on example, not the actually library. Thanks
